src/wayland/handlers/xdg_shell/mod.rs

```diff
@@ -75,52 +78,75 @@ impl XdgShellHandler for State {
     fn grab(&mut self, surface: PopupSurface, seat: WlSeat, serial: Serial) {
         let seat = Seat::from_resource(&seat).unwrap();
         let kind = PopupKind::Xdg(surface);
-        let maybe_root = find_popup_root_surface(&kind)
-            .ok()
-            .and_then(|root| self.common.shell.read().element_for_surface(&root).cloned());
+        let maybe_root = find_popup_root_surface(&kind).ok();
+        if maybe_root.is_none() {
+            tracing::warn!("No root surface found for popup grab.");
+            return;
+        }
+        let maybe_root: Option<KeyboardFocusTarget> = maybe_root.and_then(|root| {
+            let shell = self.common.shell.read();
+            shell
+                .element_for_surface(&root)
+                .map(|r| r.clone().into())
+                .or_else(|| {
+                    shell.outputs().find_map(|o| {
+                        layer_map_for_output(o).layers().find_map(|layer| {
```

Comment:
We do have `layer_for_surface`, which is probably more appropriate here.

So something like: `layer_map_for_output(o).layer_for_surface(root, WindowSurfaceType::ALL).cloned().map(KeyboardFocusTarget::LayerSurface)`, which would then also be able to find subsurfaces.

Reply:
Ok, I'll try that, thanks!